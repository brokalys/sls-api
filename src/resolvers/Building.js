import { UserInputError } from 'apollo-server-lambda';
import validationSchema from './query/properties/validation';

export default {
  id: (building) => building.id,
  bounds: (building) => {
    const bounds =
      typeof building.bounds === 'string'
        ? JSON.parse(building.bounds)
        : building.bounds;
    return bounds[0].map(({ x, y }) => [x, y].join(' ')).join(', ');
  },
  properties: (building, input, { dataSources }) => {
    const validator = validationSchema.validate(input);

    // Validate input
    if (validator.error) {
      throw new UserInputError('Input validation failed', {
        details: validator.error.details,
      });
    }

    return dataSources.properties
      .loadByBuildingId(building.id, validator.value.filter)
      .then((results) => results.map(({ id }) => id));
  },
  vzd: (building) => building,
};
