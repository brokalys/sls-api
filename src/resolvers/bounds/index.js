/* istanbul ignore file */
export { default } from './bounds';
