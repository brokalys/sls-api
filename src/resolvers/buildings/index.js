/* istanbul ignore file */
export { default } from './buildings';
