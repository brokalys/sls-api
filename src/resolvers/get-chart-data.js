import { UserInputError } from 'apollo-server-lambda';
import Moment from 'moment';
import numbers from 'numbers';

import { getRegionsData } from './get-regions';

import cache from '../lib/cache';
import mysql from '../lib/db';
import Repository from '../lib/repository';

async function getChartData(parent, { category, type, date }) {
  const month = Moment(date, 'YYYY-MM-DD');

  // Validate
  if (month.isBefore('2018-01-01')) {
    throw new UserInputError('Date must be after 2018-01-01');
  }

  if (month.isAfter(Moment().startOf('month'))) {
    throw new UserInputError(
      'Date must not be bigger than start of current month',
    );
  }

  const start = month.clone().startOf('month').toISOString();
  const end = month.clone().endOf('month').toISOString();

  const data = await cache.run(
    'getChartData.dataRetrieval',
    { category, type, start, end },
    Repository.getRawChartData,
  );

  const medianPrice = numbers.statistic.median(
    data.map(({ price_per_sqm }) => price_per_sqm),
  );

  return {
    price_per_sqm: (medianPrice || 0).toFixed(2),
    count: data.length,
  };
}

export default getChartData;
