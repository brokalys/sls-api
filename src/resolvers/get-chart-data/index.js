import fn from './get-chart-data';
export default fn;
