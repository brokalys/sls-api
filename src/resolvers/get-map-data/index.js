import fn from './get-map-data';
export default fn;
