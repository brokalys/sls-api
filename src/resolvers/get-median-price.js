import { UserInputError } from 'apollo-server-lambda';
import moment from 'moment';
import numbers from 'numbers';

import cache from '../lib/cache';
import Repository from '../lib/repository';
import Joi from '../lib/validator';

// Validation schema
const validationSchema = Joi.object().keys({
  category: Joi.string().allow('APARTMENT', 'HOUSE', 'LAND'),
  type: Joi.string().allow('SELL', 'RENT'),
  start_date: Joi.date()
    .iso()
    .max(moment().utc().startOf('month').subtract(1, 'day'))
    .required(),
  region: Joi.string().required().polygonV2(),
});

async function getMedianPrice(parent, args) {
  const validator = validationSchema.validate(args);

  // Validate input
  if (validator.error) {
    throw new UserInputError('Input validation failed', {
      details: validator.error.details,
    });
  }

  const start = moment(args.start_date).utc().startOf('month');
  const end = start.clone().endOf('month');

  return await cache.run(
    'getMedianPrice',
    {
      category: args.category,
      type: args.type,
      start: start.format('YYYY-MM-DD'),
      end: end.format('YYYY-MM-DD'),
      region: args.region,
    },
    getData,
  );
}

async function getData(input) {
  const prices = await Repository.getPricesInRegion(input);

  return {
    price: Math.ceil(numbers.statistic.median(prices)) || null,
  };
}

export default getMedianPrice;
