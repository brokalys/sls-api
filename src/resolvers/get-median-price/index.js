import fn from './get-median-price';
export default fn;
