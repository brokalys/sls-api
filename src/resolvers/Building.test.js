import { UserInputError } from 'apollo-server-lambda';
import PropertiesDataSource from 'data-sources/properties';
import resolvers from './Building';

jest.mock('data-sources/properties');

const building = {
  id: 1,
  bounds: [
    [
      { x: 56.992294, y: 24.136619 },
      { x: 56.976394, y: 23.99579 },
      { x: 56.924904, y: 24.005336 },
      { x: 56.889288, y: 24.108467 },
      { x: 56.932211, y: 24.291935 },
      { x: 56.996502, y: 24.245176 },
      { x: 56.992294, y: 24.136619 },
    ],
  ],
};

describe('Building', () => {
  describe('id', () => {
    it('returns id value', () => {
      const output = resolvers.id(building);
      expect(output).toEqual(building.id);
    });
  });

  describe('bounds', () => {
    it('returns bounds value', () => {
      const output = resolvers.bounds(building);
      expect(output).toMatchSnapshot();
    });

    it('JSON decodes the value if it is a string', () => {
      const output = resolvers.bounds({
        ...building,
        bounds: JSON.stringify(building.bounds),
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('properties', () => {
    let dataSources;

    beforeEach(() => {
      dataSources = {
        properties: PropertiesDataSource,
      };
    });

    it('retrieves property ids for the given building', async () => {
      PropertiesDataSource.loadByBuildingId.mockResolvedValueOnce([
        { id: 1 },
        { id: 2 },
        { id: 3 },
      ]);

      const output = await resolvers.properties(building, {}, { dataSources });

      expect(output).toEqual([1, 2, 3]);
    });

    it('fails validating the provided filter', async () => {
      expect(() => {
        resolvers.properties(building, { filter: 'wrong' }, { dataSources });
      }).toThrowError(UserInputError);
    });
  });
});
