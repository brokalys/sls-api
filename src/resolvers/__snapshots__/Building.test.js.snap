// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building bounds JSON decodes the value if it is a string 1`] = `"56.992294 24.136619, 56.976394 23.99579, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176, 56.992294 24.136619"`;

exports[`Building bounds returns bounds value 1`] = `"56.992294 24.136619, 56.976394 23.99579, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176, 56.992294 24.136619"`;
